### ISSUE TITLE
Different Gradients with Two API

### ISSUE THREAD

**Issue by chao1224** [pytorch/pytorch#2696]
Hi there,

I tried to get the gradient w.r.t. the neural network weights.

AFAIK, there are two APIs.
1. The [auto grad function](https://github.com/pytorch/pytorch/blob/master/torch/autograd/__init__.py#L101)
2. The [attribute in class Variable](https://github.com/pytorch/pytorch/blob/master/torch/autograd/variable.py#L26-L28)

Now I build a toy model, and layer is `4 -> 10 -> 20 -> 5 -> 1`, and I hope to get the gradient w.r.t. weight parameter in each layer.

What I get for the output is that, only the gradients in first layer `4*10` are the same using two APIs, all the remaining layers are different. Am I missing sth?

Any help would be helpful.

**Comment by soumith** [comment]
can you post a snippet for this toy model / validating gradients. will look into it.

**Comment by chao1224** [comment]
@soumith Sure, I put them into a [gist](https://gist.github.com/chao1224/1f07666d44a4c2d99cd5061b79d49729)

Thank you in advance.

**Comment by fmassa** [comment]
@chao1224 Don't you need to zero the gradients before computing `autograd.grad`, or else they can be accumulated?

**Comment by chao1224** [comment]
@fmassa Thanks for pointing out. Now this works fine for me.
Appreciate your help.
